// typescript-json/lib/transformers/CallExpressionTransformer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallExpressionTransformer = void 0;
var path_1 = __importDefault(require("path"));
var AssertCloneTransformer_1 = require("./features/cloners/AssertCloneTransformer");
var CloneTransformer_1 = require("./features/cloners/CloneTransformer");
var CreateAssertCloneTransformer_1 = require("./features/cloners/CreateAssertCloneTransformer");
var CreateCloneTransformer_1 = require("./features/cloners/CreateCloneTransformer");
var CreateIsCloneTransformer_1 = require("./features/cloners/CreateIsCloneTransformer");
var IsCloneTransformer_1 = require("./features/cloners/IsCloneTransformer");
var ApplicationTransformer_1 = require("./features/miscellaneous/ApplicationTransformer");
var CreateInstanceTransformer_1 = require("./features/miscellaneous/CreateInstanceTransformer");
var AssertStringifyTransformer_1 = require("./features/stringifiers/AssertStringifyTransformer");
var CreateAssertStringifyTransformer_1 = require("./features/stringifiers/CreateAssertStringifyTransformer");
var CreateIsStringifyTransformer_1 = require("./features/stringifiers/CreateIsStringifyTransformer");
var CreateStringifyTransformer_1 = require("./features/stringifiers/CreateStringifyTransformer");
var IsStringifyTransformer_1 = require("./features/stringifiers/IsStringifyTransformer");
var StringifyTransformer_1 = require("./features/stringifiers/StringifyTransformer");
var AssertTransformer_1 = require("./features/validators/AssertTransformer");
var CreateAssertTransformer_1 = require("./features/validators/CreateAssertTransformer");
var CreateIsTransformer_1 = require("./features/validators/CreateIsTransformer");
var CreateValidateTransformer_1 = require("./features/validators/CreateValidateTransformer");
var IsTransformer_1 = require("./features/validators/IsTransformer");
var ValidateTransformer_1 = require("./features/validators/ValidateTransformer");
var CallExpressionTransformer;
(function (CallExpressionTransformer) {
    function transform(project, expression) {
        var _a;
        var declaration = (_a = project.checker.getResolvedSignature(expression)) === null || _a === void 0 ? void 0 : _a.declaration;
        if (!declaration)
            return expression;
        var file = path_1.default.resolve(declaration.getSourceFile().fileName);
        if (file !== LIB_PATH && file !== SRC_PATH)
            return expression;
        var name = project.checker.getTypeAtLocation(declaration).symbol.name;
        var functor = FUNCTORS[name];
        if (functor === undefined)
            return expression;
        return functor()(project, expression.expression, expression);
    }
    CallExpressionTransformer.transform = transform;
})(CallExpressionTransformer = exports.CallExpressionTransformer || (exports.CallExpressionTransformer = {}));
var LIB_PATH = path_1.default.resolve(path_1.default.join(__dirname, "..", "module.d.ts"));
var SRC_PATH = path_1.default.resolve(path_1.default.join(__dirname, "..", "module.ts"));
var FUNCTORS = {
    assert: function () { return AssertTransformer_1.AssertTransformer.transform(false); },
    assertType: function () { return AssertTransformer_1.AssertTransformer.transform(false); },
    is: function () { return IsTransformer_1.IsTransformer.transform(false); },
    validate: function () { return ValidateTransformer_1.ValidateTransformer.transform(false); },
    assertEquals: function () { return AssertTransformer_1.AssertTransformer.transform(true); },
    equals: function () { return IsTransformer_1.IsTransformer.transform(true); },
    validateEquals: function () { return ValidateTransformer_1.ValidateTransformer.transform(true); },
    stringify: function () { return StringifyTransformer_1.StringifyTransformer.transform; },
    assertStringify: function () { return AssertStringifyTransformer_1.AssertStringifyTransformer.transform; },
    isStringify: function () { return IsStringifyTransformer_1.IsStringifyTransformer.transform; },
    clone: function () { return CloneTransformer_1.CloneTransformer.transform; },
    isClone: function () { return IsCloneTransformer_1.IsCloneTransformer.transform; },
    assertClone: function () { return AssertCloneTransformer_1.AssertCloneTransformer.transform; },
    application: function () { return ApplicationTransformer_1.ApplicationTransformer.transform; },
    createAssert: function () { return CreateAssertTransformer_1.CreateAssertTransformer.transform(false); },
    createAssertType: function () { return CreateAssertTransformer_1.CreateAssertTransformer.transform(false); },
    createIs: function () { return CreateIsTransformer_1.CreateIsTransformer.transform(false); },
    createValidate: function () { return CreateValidateTransformer_1.CreateValidateTransformer.transform(false); },
    createAssertEquals: function () { return CreateAssertTransformer_1.CreateAssertTransformer.transform(true); },
    createEquals: function () { return CreateIsTransformer_1.CreateIsTransformer.transform(true); },
    createValidateEquals: function () { return CreateValidateTransformer_1.CreateValidateTransformer.transform(true); },
    createStringify: function () { return CreateStringifyTransformer_1.CreateStringifyTransformer.transform; },
    createAssertStringify: function () { return CreateAssertStringifyTransformer_1.CreateAssertStringifyTransformer.transform; },
    createIsStringify: function () { return CreateIsStringifyTransformer_1.CreateIsStringifyTransformer.transform; },
    createClone: function () { return CreateCloneTransformer_1.CreateCloneTransformer.transform; },
    createIsClone: function () { return CreateIsCloneTransformer_1.CreateIsCloneTransformer.transform; },
    createAssertClone: function () { return CreateAssertCloneTransformer_1.CreateAssertCloneTransformer.transform; },
    createObject: function () { return CreateInstanceTransformer_1.CreateInstanceTransformer.transform; },
};
